File: packages/python/src/openskp/transforms.py
"""3-D matrix and point transforms used during SKP parsing.

SketchUp stores instance placement as 4×4 affine transformation matrices
in column-major order.  This module provides helpers to multiply those
matrices, apply them to points, and convert between coordinate systems
(SketchUp uses inches / Z-up; many renderers expect metres / Y-up).

All matrices are stored as flat 16-element lists in **column-major** order
matching the layout found in the binary file.
"""

from __future__ import annotations

from typing import List, Tuple

# ── Constants ─────────────────────────────────────────────────────────────

INCHES_TO_METRES: float = 0.0254
"""Conversion factor from SketchUp inches to SI metres."""

IDENTITY_MATRIX: List[float] = [
    1.0, 0.0, 0.0, 0.0,
    0.0, 1.0, 0.0, 0.0,
    0.0, 0.0, 1.0, 0.0,
    0.0, 0.0, 0.0, 1.0,
]
"""4×4 identity matrix in column-major order."""


# ── Matrix helpers ────────────────────────────────────────────────────────


def multiply_matrices(a: List[float], b: List[float]) -> List[float]:
    """Multiply two 4×4 column-major matrices.

    Args:
        a: First matrix (16 floats, column-major).
        b: Second matrix (16 floats, column-major).

    Returns:
        The product ``a × b`` as a 16-element column-major list.
    """
    result: List[float] = [0.0] * 16
    for row in range(4):
        for col in range(4):
            s = 0.0
            for k in range(4):
                # column-major indexing: element (row, col) at index col*4+row
                s += a[k * 4 + row] * b[col * 4 + k]
            result[col * 4 + row] = s
    return result


def transform_point(
    matrix: List[float],
    x: float,
    y: float,
    z: float,
) -> Tuple[float, float, float]:
    """Apply a 4×4 affine transform to a 3-D point.

    Args:
        matrix: 16-element column-major transformation matrix.
        x: X coordinate.
        y: Y coordinate.
        z: Z coordinate.

    Returns:
        The transformed ``(x', y', z')`` tuple.
    """
    # M is column-major: column c starts at index c*4
    tx = matrix[0] * x + matrix[4] * y + matrix[8] * z + matrix[12]
    ty = matrix[1] * x + matrix[5] * y + matrix[9] * z + matrix[13]
    tz = matrix[2] * x + matrix[6] * y + matrix[10] * z + matrix[14]
    return (tx, ty, tz)


def z_up_to_y_up(x: float, y: float, z: float) -> Tuple[float, float, float]:
    """Convert from SketchUp's Z-up to a Y-up coordinate system.

    The mapping is ``(x, y, z) → (x, z, -y)``.

    Args:
        x: X coordinate.
        y: Y coordinate (originally "depth" in SketchUp).
        z: Z coordinate (originally "up" in SketchUp).

    Returns:
        Re-oriented ``(x', y', z')`` tuple.
    """
    return (x, z, -y)


def inches_to_metres(x: float, y: float, z: float) -> Tuple[float, float, float]:
    """Convert a point from inches to metres.

    Args:
        x: X in inches.
        y: Y in inches.
        z: Z in inches.

    Returns:
        ``(x, y, z)`` in metres.
    """
    return (x * INCHES_TO_METRES, y * INCHES_TO_METRES, z * INCHES_TO_METRES)


def decompose_translation(matrix: List[float]) -> Tuple[float, float, float]:
    """Extract the translation component from a 4×4 column-major matrix.

    Args:
        matrix: 16-element column-major transformation matrix.

    Returns:
        ``(tx, ty, tz)`` translation vector.
    """
    return (matrix[12], matrix[13], matrix[14])


def is_identity(matrix: List[float], tol: float = 1e-9) -> bool:
    """Check whether a 4×4 matrix is (approximately) the identity.

    Args:
        matrix: 16-element column-major transformation matrix.
        tol: Absolute tolerance for floating-point comparison.

    Returns:
        ``True`` if every element matches the identity within *tol*.
    """
    for i in range(16):
        expected = 1.0 if (i % 5 == 0) else 0.0
        if abs(matrix[i] - expected) > tol:
            return False
    return True

File: packages/dotnet/OpenSkp/Transforms.cs
using System;
using System.Collections.Generic;

namespace OpenSkp
{
    /// <summary>3D point/matrix helpers for scene baking. Ported from
    /// Python's _core.py (transform_point / multiply_matrices) - matrices
    /// are SketchUp's 13-element [3x3 rotation/scale | translation | 1.0]
    /// layout unless noted.</summary>
    internal static class Transforms
    {
        public static (double X, double Y, double Z) TransformPoint(double[] matrix, (double X, double Y, double Z) point)
        {
            if (matrix == null || matrix.Length < 12)
            {
                return point;
            }
            double x = point.X, y = point.Y, z = point.Z;
            double tx = matrix[0] * x + matrix[1] * y + matrix[2] * z + matrix[9];
            double ty = matrix[3] * x + matrix[4] * y + matrix[5] * z + matrix[10];
            double tz = matrix[6] * x + matrix[7] * y + matrix[8] * z + matrix[11];
            return (tx, ty, tz);
        }

        public static List<double> MultiplyMatrices(List<double> parent, List<double> child)
        {
            if (parent == null || parent.Count == 0) return child;
            if (child == null || child.Count == 0) return parent;

            double[] p = Pad13(parent);
            double[] c = Pad13(child);

            double[] pR0 = { p[0], p[1], p[2], p[9] };
            double[] pR1 = { p[3], p[4], p[5], p[10] };
            double[] pR2 = { p[6], p[7], p[8], p[11] };

            double[] cC0 = { c[0], c[3], c[6], 0 };
            double[] cC1 = { c[1], c[4], c[7], 0 };
            double[] cC2 = { c[2], c[5], c[8], 0 };
            double[] cC3 = { c[9], c[10], c[11], 1 };

            static double Dot(double[] row, double[] col) =>
                row[0] * col[0] + row[1] * col[1] + row[2] * col[2] + row[3] * col[3];

            var outArr = new double[13];
            outArr[0] = Dot(pR0, cC0);
            outArr[1] = Dot(pR0, cC1);
            outArr[2] = Dot(pR0, cC2);
            outArr[3] = Dot(pR1, cC0);
            outArr[4] = Dot(pR1, cC1);
            outArr[5] = Dot(pR1, cC2);
            outArr[6] = Dot(pR2, cC0);
            outArr[7] = Dot(pR2, cC1);
            outArr[8] = Dot(pR2, cC2);
            outArr[9] = Dot(pR0, cC3);
            outArr[10] = Dot(pR1, cC3);
            outArr[11] = Dot(pR2, cC3);
            outArr[12] = p[12] * c[12];
            return new List<double>(outArr);
        }

        private static double[] Pad13(List<double> m)
        {
            var arr = new double[13];
            for (int i = 0; i < 13; i++)
            {
                arr[i] = i < m.Count ? m[i] : 0.0;
            }
            if (m.Count < 13) arr[12] = 1.0;
            return arr;
        }
    }
}
